language/compiler/ir_to_bytecode/syntax/src/ast.rs

```diff
@@ -143,13 +145,30 @@ pub enum StructDefinitionFields {
 #[derive(Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Clone)]
 pub struct FunctionName(String);
 
+/// The kind of a type
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum Kind {
+    All,
+    Resource,
+    Unrestricted,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct KindConstraint(pub String, pub Kind);
+
+/// A list of kind constraints
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct KindConstraints(pub Vec<KindConstraint>);
+
 /// The signature of a function
 #[derive(PartialEq, Debug, Clone)]
 pub struct FunctionSignature {
     /// Possibly-empty list of (formal name, formal type) pairs. Names are unique.
     pub formals: Vec<(Var, Type)>,
     /// Optional return types
     pub return_type: Vec<Type>,
+    /// Kind constraints of the type parameters
+    pub kind_constraints: KindConstraints,
```

Comment:
same as above

Reply:
E.g., it looks a bit funky that we have `Vec`'s for `return_type` and `formals` but not this.